src/core/io.c

```diff
@@ -127,35 +129,80 @@ static Janet cfun_io_popen(int32_t argc, Janet *argv) {
 #define popen _popen
 #endif
     FILE *f = popen((const char *)fname, (const char *)fmode);
-    if (!f) {
+    
+    if(NULL == f) {
         return janet_wrap_nil();
+    } else {
+        char * buffer = NULL;
+        
+        if(argc == 3) {
+            int bufsiz = janet_getinteger(argv, 2);
+            
+            if(bufsiz <= 0 && setvbuf(f, NULL, _IONBF, 0))
+                return janet_wrap_nil();
+            
+            buffer = malloc(bufsiz);
+            
+            if(NULL == buffer) {
+                janet_panic("memory exhausted");
+            }
+            
+            if(setvbuf(f, buffer, _IOFBF, bufsiz))
+                return janet_wrap_nil();
+        }
+        
+        return makef(f, flags, buffer);
     }
-    return makef(f, flags);
 }
 #endif
 
 static Janet cfun_io_fopen(int32_t argc, Janet *argv) {
-    janet_arity(argc, 1, 2);
+    janet_arity(argc, 1, 3);
     const uint8_t *fname = janet_getstring(argv, 0);
     const uint8_t *fmode;
     int flags;
-    if (argc == 2) {
+    
+    if (argc >= 2) {
         fmode = janet_getkeyword(argv, 1);
         flags = checkflags(fmode);
     } else {
         fmode = (const uint8_t *)"r";
         flags = JANET_FILE_READ;
     }
+    
     FILE *f = fopen((const char *)fname, (const char *)fmode);
-    return f ? makef(f, flags) : janet_wrap_nil();
+    
+    if(NULL == f) {
+        return janet_wrap_nil();
+    } else {
+        char * buffer = NULL;
+        
+        if(argc == 3) {
+            int bufsiz = janet_getinteger(argv, 2);
+            
+            if(bufsiz <= 0 && setvbuf(f, NULL, _IONBF, 0))
+                return janet_wrap_nil();
+            
+            buffer = malloc(bufsiz);
+            
+            if(NULL == buffer) {
+                janet_panic("memory exhausted");
```

Comment:
Memory exhaustion is should not be handled like this, there is a specific out of memory define you should use.

Reply:
Done.